### PR TITLE
[feature] Cookie consent (#1055)

### PR DESCRIPTION
**Pull request by donrestarone**
# Cookie consent for compliance
## Demo
https://user-images.githubusercontent.com/50227291/186684043-b502d319-2865-4121-9d84-4c7789c17496.mov

Customizable cookies consent banner

https://user-images.githubusercontent.com/50227291/186684374-a7db4ecc-2c1b-4b78-98ba-f9ff680efd58.mov

### CODE REVIEW

File: app/views/cookies/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+<%= turbo_frame_tag :cookies_modal do %>
+  <% if cookies[:cookies_accepted].nil? %>
+    <%= Subdomain.current.cookies_consent_ui.html_safe %>
```
[the remark this answers is not in the document]
Comment (donrestarone) [reply app/views/cookies/index.html.erb:3]: @Pralish does this alert need to be addressed?

**Comment by Ayon95** [comment]
Should I customize the cookie consent popup from the frontend, or make a commit to this branch containing the cookie consent markup?

**Comment by donrestarone** [comment]
@Ayon95 yes and no.

Let me explain-- the code here for the pop up is the default styling and structure. This is what we will ship out of the box. So you will need to create a UI that looks on-brand for Violet Rails and has 2 buttons. Please coordinate with @Pralish on this but I recommend opening a PR against this branch (using this as your base branch, so checkout from here not master) with your changes for the default markup.


That said, once deployed you will immediately go to the Nikean Admin and customize it such that it aligns with Nikean's brand design/language.

**Comment by Ayon95** [comment]
Is it necessary to open a PR against this branch with my changes since I will have to customize the markup and styling for Nikean from admin?

**Comment by donrestarone** [comment]
@Ayon95 to be clear, you need to do 2 things.

First, submit a default UI for the cookie consent on this branch. 

Second, Once deployed you will customize the pop up for clients on a per site basis.

**Comment by Ayon95** [comment]
Okay, thanks for clarifying. I'll have to take a look at it later. I'm out at the moment.

**Comment by donrestarone** [comment]
# Acceptance criteria 📋 

1. General regression pass: API Actions ⌛ 
2. General regression pass: API Resource Tracking - visitors and users ⌛  
3. General regression pass: Users Tracking ⌛  
4. General regression pass: Analytics - visitors and users tracking ⌛  
5. General regression pass: Ahoy JS events ⌛  
6. Feature validation: Ahoy JS events/tracking fails gracefully if the user has declined cookies
7. Feature validation: Ahoy JS events/tracking does not occur if the user has declined cookies
8. Feature validation: visits are not logged if the user has declined cookies - (as a visitor and signed in user)

**Comment by donrestarone** [comment]
❌ blockers (WIP ⌛ ) 

1. Redactor is broken on first render (refreshing the page fixes it)-- this should not be the case:
Screen Shot 2022-08-28 at 2 01 51 PM

Related to the above, snippet renderer is broken as well: 


Screen Shot 2022-08-28 at 2 04 31 PM


2. Default banner should have a solid background for legibility over various interfaces 
Screen Shot 2022-08-28 at 1 50 32 PM

**Comment by Ayon95** [comment]
Okay, I'll implement this layout then. Should I use white for the background color?

**Comment by donrestarone** [comment]
@Ayon95 yup, feel free to use white or black depending on your best judgement. 

The key is to show user (like in the OMX example I illustrated above) that this pop up is not dismissible without making a choice. Nothing under it should be clickable and everything around it should be grayed out

**Comment by Ayon95** [comment]
Does this look fine?

Mobile:
cookies-consent-mobile

Desktop:
cookies-consent-desktop

**Comment by donrestarone** [comment]
@Ayon95 looks good! Nothing is clickable in that Gray'd out area right?

**Comment by Ayon95** [comment]
Yeah, nothing underneath the gray area can be clicked.

**Comment by donrestarone** [comment]
@Ayon95 that works. Btw before you open the PR you should git fetch and check out your change branch from the top of this branch (since I merged your last change in already)

**Comment by donrestarone** [comment]
@Ayon95 your client-side changes look good on [Heroku](http://violetrails.herokuapp.com) 🌏 

Screen Shot 2022-08-29 at 2 24 25 PM


@Pralish you're up next with [the server changes](https://github.com/restarone/violet_rails/pull/1059#issuecomment-1229521191) 🤖

**Comment by donrestarone** [comment]
TODO
@Pralish we will need to document how to customize the popup: 

Screen Shot 2022-08-29 at 2 35 13 PM

**Comment by Pralish** [comment]
@donrestarone , I have addressed the issue. Can we have another round of alpha testing.

**Comment by donrestarone** [comment]
# Acceptance criteria 📋 

1. General regression pass: API Actions ⌛    
2. General regression pass: API Resource Tracking - visitors and users ⌛ 
3. General regression pass: Users Tracking⌛ 
4. General regression pass: Analytics - visitors and users tracking ⌛ 
5. General regression pass: Ahoy JS events ⌛    
6. General regression pass: visit/events are tracked after consent ⌛ *pending fix* 
7. Feature validation: Ahoy JS events/tracking fails gracefully if the user has declined cookies ⌛  
8.  Feature validation: Cookie consent is only shown once ⌛ 
9. Feature validation: Cookie consent is shown in forum ⌛ 
10. Feature validation: Cookie consent is shown in blog ⌛    
11. Feature validation: Cookie consent is shown in web ⌛    
12. Feature validation: Cookie consent is shown in admin ⌛ 
13. Feature validation: Ahoy JS events/tracking does not occur if the user has declined cookies ⌛   
14. Feature validation: visits are not logged if the user has declined cookies - (as a visitor and signed in user)⌛  
15. Consent popup is not shown if tracking is not enabled ⌛ 
16. General regression pass: Recaptcha V2 and V3 forms ⌛ *pending `staging`*

File: app/views/shared/_cookies_consent_ui.html.erb [hunk 2]
```diff
@@ -0,0 +1 @@
+<%= Subdomain.current.cookies_consent_ui.html_safe if show_cookies_consent_banner? %>
```
[the remark this answers is not in the document]
Comment (donrestarone) [reply app/views/shared/_cookies_consent_ui.html.erb:1]: @Pralish do we need to address this finding?

[the remark this answers is not in the document]
Comment (Pralish) [reply app/views/shared/_cookies_consent_ui.html.erb]: @donrestarone , This has been addressed. Thanks

**Comment by donrestarone** [comment]
thanks @Pralish @Ayon95 ! 

⚠️ Temporary canary deployment to `staging`

**Comment by donrestarone** [comment]
The default consent popup isn't perfect, here are some examples where it's broken 'out of the box'


Screen Shot 2022-08-31 at 9 10 27 PM
Screen Shot 2022-08-31 at 9 10 11 PM

When this goes into production we should be ready to address issues as they come up.

**Pull request merged**